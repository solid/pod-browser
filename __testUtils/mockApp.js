/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
