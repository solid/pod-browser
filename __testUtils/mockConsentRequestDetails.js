/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

async function getConsentRequestDetails() {
  return {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
      "https://consent.pod.inrupt.com/credentials/v1",
    ],
    id: "https://consent.pod.inrupt.com/vc/53973727-f4d0-9e8dbdc041fd",
    type: ["VerifiableCredential", "SolidCredential", "SolidConsentRequest"],
    issuer: "https://consent.pod.inrupt.com",
    issuanceDate: "2021-05-26T16:40:03Z",
    expirationDate: "2021-06-23T16:40:03Z",
    credentialSubject: {
      id: "https://pod.inrupt.com/bob/profile/card#me",
      inbox: "https://pod.inrupt.com/bob/inbox/",
      hasConsent: {
        mode: ["Read", "Write"],
        hasStatus: "ConsentStatusRequested",
        forPersonalData: "https://pod.inrupt.com/alice/private/data",
        forPurpose: "https://example.com/SomeSpecificPurpose",
      },
    },
    proof: {
      created: "2021-05-26T16:40:03.009Z",
      proofPurpose: "assertionMethod",
      proofValue: "eqp8h_kL1DwJCpn65z-d1Arnysx6b11...jb8j0MxUCc1uDQ",
      type: "Ed25519Signature2020",
      verificationMethod: "https://consent.pod.inrupt.com/key/396f686b4ec",
    },
  };
}

export default getConsentRequestDetails;
