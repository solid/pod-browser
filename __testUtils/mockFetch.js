/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export default function mockFetch(responses = {}) {
  return jest.fn().mockImplementation((url, init) => {
    const responseFn = responses[url.toString()];
    const responseObject = responseFn;
    const method = init?.method || "GET";
    let response;
    if (responseFn && typeof responseObject[method] !== "undefined") {
      response = responseObject[method](url, init);
    } else if (responseFn) {
      response = responseFn(url, init);
    } else {
      throw new Error(`URL (${url}) not mocked properly`);
    }
    Object.defineProperty(response, "url", { value: url });
    return Promise.resolve(response);
  });
}

export function mockSolidClientRequest(body, method = "GET", headers = {}) {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status, body = "") {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
