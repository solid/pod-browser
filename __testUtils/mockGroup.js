/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  removeUrl,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { foaf, rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { getBaseUrl } from "../src/solidClientHelpers/resource";

export function addMembersToMockedGroup(group, agentUrls) {
  const dataset = chain(
    group.dataset,
    ...agentUrls.map((agentUrl) => (d) =>
      setThing(d, setUrl(group.thing, vcard.hasMember, agentUrl))
    )
  );

  return {
    dataset,
    thing: getThing(dataset, asUrl(group.thing)),
  };
}

export function removeMembersFromMockedGroup(group, agentUrls) {
  const dataset = chain(
    group.dataset,
    ...agentUrls.map((agentUrl) => (d) =>
      setThing(d, removeUrl(group.thing, vcard.hasMember, agentUrl))
    )
  );

  return {
    dataset,
    thing: getThing(dataset, asUrl(group.thing)),
  };
}

export function mockGroupThing(name, url, { description, members } = {}) {
  return chain(
    mockThingFrom(url),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, foaf.name, name),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    ),
    ...[
      description ? (t) => setStringNoLocale(t, vcard.note, description) : null,
    ].filter((fn) => !!fn)
  );
}

export default function mockGroup(name, url, options = {}) {
  const thing = mockGroupThing(name, url, options);
  const dataset = setThing(mockSolidDatasetFrom(getBaseUrl(url)), thing);
  return { dataset, thing };
}
