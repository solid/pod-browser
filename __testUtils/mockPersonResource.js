/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  asUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { vcard, foaf, rdf, space } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { packageProfile } from "../src/solidClientHelpers/profile";
import { vcardExtras } from "../src/addressBook";

const VCARD_WEBID_PREDICATE = vcardExtras("WebId");

export function mockWebIdNode(
  webId,
  nodeUrl = "https://example.org/contacts/Person/1234/index.ttl#4567"
) {
  const webIdNode = chain(
    mockThingFrom(nodeUrl),
    (t) => addUrl(t, rdf.type, VCARD_WEBID_PREDICATE),
    (t) => addUrl(t, vcard.value, webId)
  );
  const url = asUrl(webIdNode);
  return { webIdNode, webIdNodeUrl: url };
}

export const aliceProfileUrl = "http://alice.example.com/alice";
export const aliceWebIdUrl = "http://alice.example.com/alice#me";
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://alice.example.com/alice.jpg";
export const alicePodRoot = "http://alice.example.com/";
export const aliceAlternativeProfileUrl = "https://alice2.example.org/card";
export const aliceAlternativeWebIdUrl = "https://alice2.example.org/card#me";

export function mockPersonThingAlice(...operations) {
  return chain(
    mockThingFrom(aliceWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, vcard.url, aliceAlternativeWebIdUrl),
    (t) => addUrl(t, space.storage, alicePodRoot),
    ...operations
  );
}

export function mockPersonDatasetAlice(...operations) {
  return setThing(
    mockSolidDatasetFrom(aliceWebIdUrl),
    mockPersonThingAlice(...operations)
  );
}

export function mockProfileAlice(...operations) {
  return packageProfile(aliceWebIdUrl, mockPersonDatasetAlice(...operations));
}

export const bobProfileUrl = "http://bob.example.com/bob";
export const bobWebIdUrl = "http://bob.example.com/bob#me";
export const bobName = "Bob";
export const bobNick = "B";
export const bobPodRoot = "http://bob.example.com/";
export const bobAlternateProfileUrl = "https://bob2.example.org/card";
export const bobAlternateWebIdUrl = "https://bob2.example.org/card#me";

export function mockPersonThingBob() {
  return chain(
    mockThingFrom(bobWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, bobName),
    (t) => addStringNoLocale(t, foaf.nick, bobNick),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, foaf.openid, bobAlternateWebIdUrl),
    (t) => addUrl(t, space.storage, bobPodRoot)
  );
}

export function mockPersonDatasetBob() {
  return setThing(mockSolidDatasetFrom(bobWebIdUrl), mockPersonThingBob());
}

export function mockProfileBob() {
  return packageProfile(bobWebIdUrl, mockPersonDatasetBob());
}
