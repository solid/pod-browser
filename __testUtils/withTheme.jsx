/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
