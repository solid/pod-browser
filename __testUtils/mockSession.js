/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import mockFetch from "./mockFetch";
import mockResponse from "./mockResponse";

import profileTurtle from "./mocks/profile.ttl";
import storageTurtle from "./mocks/podRoot.ttl";
import storageAclTurtle from "./mocks/storageAcl.ttl";

export const webIdUrl = "http://example.com/webId#me";
export const storageUrl = "http://example.com/";
export const storageAclUrl = "http://example.com/.acl";
export const anotherUsersStorageUrl = "http://anotheruser.com/";
export const anotherUsersStorageAclUrl = "http://anotheruser.com/.acl";

export { profileTurtle };

export default function mockSession(options = {}) {
  // this instance is a normal, authenticated session
  // you can override the properties with what you need, but if you repeat something a lot
  // consider naming it and create a reusable mock function
  const { webId, ...restOptions } = options;
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webId || webIdUrl]: () =>
        mockResponse(200, profileTurtle, { "Content-Type": "text/turtle" }),
      [storageUrl]: () =>
        mockResponse(200, storageTurtle, {
          Link: `<${storageAclUrl}>; rel="acl"`,
          "Content-Type": "text/turtle",
        }),
      [storageAclUrl]: () =>
        mockResponse(200, storageAclTurtle, { "Content-Type": "text/turtle" }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: options.webId || webIdUrl,
    }, // add more properties as needed
    ...restOptions,
  }; // add more properties as needed
}

// create other functions in this file to reflect other types of sessions we need to test with

export function mockUnauthenticatedSession() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: () => mockFetch(),
    info: {
      isLoggedIn: false,
      sessionId: null,
      webId: null,
    },
    login: jest.fn(),
  };
}

export function mockAuthenticatedSessionWithNoAccessToPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [storageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${storageAclUrl}>; rel="acl"`,
        }),
      [storageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export function mockAuthenticatedSessionWithNoAccessToAnotherUsersPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [anotherUsersStorageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${anotherUsersStorageAclUrl}>; rel="acl"`,
        }),
      [anotherUsersStorageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export const mockAuthenticatedSession =
  mockAuthenticatedSessionWithNoAccessToPod;
