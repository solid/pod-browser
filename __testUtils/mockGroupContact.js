/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  removeThing,
  removeUrl,
  setThing,
} from "@inrupt/solid-client";
import { chain } from "../src/solidClientHelpers/utils";
import { createGroupDatasetUrl } from "../src/models/contact/group";
import { vcardExtras } from "../src/addressBook";
import { mockGroupThing } from "./mockGroup";
import { getAddressBookDatasetUrl } from "../src/models/addressBook";

export function mockIndexThing(indexDataset, addressBook, groupThingUrl) {
  const addressBookThingUrl = asUrl(
    addressBook.thing,
    getAddressBookDatasetUrl(addressBook)
  );
  const addressBookThing =
    getThing(indexDataset, addressBookThingUrl) ||
    mockThingFrom(addressBookThingUrl);
  return addUrl(addressBookThing, vcardExtras("includesGroup"), groupThingUrl);
}

export function addGroupToMockedIndexDataset(
  dataset,
  addressBook,
  name,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => setThing(d, mockGroupThing(name, groupThingUrl)),
    (d) => setThing(d, mockIndexThing(d, addressBook, groupThingUrl))
  );
}

export function removeGroupFromMockedIndexDataset(
  dataset,
  addressBook,
  groupIndex,
  groupThing,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => removeThing(d, groupThing),
    (d) =>
      setThing(
        d,
        chain(
          getThing(groupIndex, asUrl(addressBook.thing)),
          (t) => removeUrl(t, vcardExtras("includesGroup"), groupThingUrl) // remove the includesGroup reference to group
        )
      )
  );
}

export default function mockGroupContact({
  addressBook,
  name,
  indexUrl = null,
  options = {},
}) {
  const { url, id } = options;
  const groupDatasetUrl = url || createGroupDatasetUrl(addressBook, id);
  const groupThingUrl = `${groupDatasetUrl}#this`;
  const groupThing = mockGroupThing(name, groupThingUrl, options);
  return {
    dataset: chain(
      mockSolidDatasetFrom(indexUrl || groupDatasetUrl),
      (d) => setThing(d, groupThing),
      (d) => setThing(d, mockIndexThing(d, addressBook, groupThingUrl))
    ),
    thing: groupThing,
  };
}
