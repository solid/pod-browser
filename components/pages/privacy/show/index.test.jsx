/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { waitFor } from "@testing-library/dom";
import * as solidClientFns from "@inrupt/solid-client";
import * as RouterFns from "next/router";
import { foaf, schema } from "rdf-namespaces";
import { renderWithTheme } from "../../../../__testUtils/withTheme";
import mockSessionContextProvider from "../../../../__testUtils/mockSessionContextProvider";
import mockSession from "../../../../__testUtils/mockSession";
import AgentPage from "./index";
import { aliceWebIdUrl } from "../../../../__testUtils/mockPersonResource";
import useFullProfile from "../../../../src/hooks/useFullProfile";

jest.mock("../../../../src/hooks/useFullProfile");
const mockedUseFullProfile = useFullProfile;

describe("Agent show page", () => {
  const mockProfileAlice = {
    names: ["Alice"],
    webId: aliceWebIdUrl,
    types: [foaf.Person],
    avatars: [],
    roles: [],
    organizations: [],
    contactInfo: {
      phones: [],
      emails: [],
    },
  };
  beforeEach(() => {
    mockedUseFullProfile.mockReturnValue(mockProfileAlice);
  });

  it("Renders the Agent show page", async () => {
    jest.spyOn(RouterFns, "useRouter").mockReturnValue({
      asPath: "/pathname/",
      replace: jest.fn(),
      query: {
        webId: "https://example.com/profile/card#me",
      },
    });

    const session = mockSession();
    const SessionProvider = mockSessionContextProvider(session);

    const { asFragment, queryAllByText } = renderWithTheme(
      <SessionProvider>
        <AgentPage type={schema.Person} />
      </SessionProvider>
    );
    await waitFor(() => {
      expect(queryAllByText("Alice")).toHaveLength(2);
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
