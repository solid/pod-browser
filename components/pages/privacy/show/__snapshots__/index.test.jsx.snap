// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Agent show page Renders the Agent show page 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-container"
  >
    <nav
      class="PodBrowser-back-to-nav"
    >
      <span
        class="PodBrowser-icon-caret-left PodBrowser-back-to-nav__caret"
      />
      <a
        class="PodBrowser-back-to-nav__link"
        href="/privacy"
      >
        Back to privacy
      </a>
    </nav>
  </div>
  <div
    class="PodBrowser-container"
  >
    <div
      class="PodBrowser-root PodBrowser-elevation1 PodBrowser-rounded"
      style="margin-top: 1em;"
    >
      <div
        class="PodBrowser-root PodBrowser-root"
      >
        <div
          class="PodBrowser-root PodBrowser-root"
        >
          <div
            class="PodBrowser-root PodBrowser-root"
          >
            <div
              class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
            >
              <div
                class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
              >
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root PodBrowser-fallback"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="PodBrowser-root PodBrowser-root"
          >
            <h3
              data-testid="profile-name-title"
            >
              <span
                class="PodBrowser-avatarText"
              >
                Alice
              </span>
              <a
                class="PodBrowser-headerLink"
                href="https://example.com/profile/card#me"
                rel="noreferrer"
                target="_blank"
              >
                https://example.com/profile/card#me
              </a>
            </h3>
          </div>
        </div>
        <hr />
        <div
          class="PodBrowser-root PodBrowser-root"
        >
          <div
            class="PodBrowser-root PodBrowser-root"
          >
            <label
              class="PodBrowser-root PodBrowser-root PodBrowser-animated"
              data-testid="profile-name-title"
            >
              Name
            </label>
            <span
              data-testid="profile-name-field"
            >
              Alice
            </span>
          </div>
          <div
            class="PodBrowser-root PodBrowser-root"
          >
            <label
              class="PodBrowser-root PodBrowser-root PodBrowser-animated"
            >
              Role
            </label>
            <span />
          </div>
          <div
            class="PodBrowser-root PodBrowser-root"
          >
            <label
              class="PodBrowser-root PodBrowser-root PodBrowser-animated"
            >
              Organization
            </label>
            <span />
          </div>
        </div>
        <div
          class="PodBrowser-root PodBrowser-root"
        >
          <label
            class="PodBrowser-root PodBrowser-root PodBrowser-animated"
          >
            Email Addresses
          </label>
          <div
            class="PodBrowser-root PodBrowser-root"
          >
            <table
              class="PodBrowser-table  PodBrowser-table--inherits PodBrowser-table"
              role="table"
            >
              <thead>
                <tr
                  role="row"
                >
                  <th
                    colspan="1"
                    role="columnheader"
                  />
                  <th
                    colspan="1"
                    role="columnheader"
                  />
                  <th
                    colspan="1"
                    role="columnheader"
                  />
                </tr>
              </thead>
              <tbody
                role="rowgroup"
              />
            </table>
          </div>
        </div>
        <div
          class="PodBrowser-root PodBrowser-root"
        >
          <label
            class="PodBrowser-root PodBrowser-root PodBrowser-animated"
          >
            Phone Numbers
          </label>
          <div
            class="PodBrowser-root PodBrowser-root"
          >
            <table
              class="PodBrowser-table  PodBrowser-table--inherits PodBrowser-table"
              role="table"
            >
              <thead>
                <tr
                  role="row"
                >
                  <th
                    colspan="1"
                    role="columnheader"
                  />
                  <th
                    colspan="1"
                    role="columnheader"
                  />
                  <th
                    colspan="1"
                    role="columnheader"
                  />
                </tr>
              </thead>
              <tbody
                role="rowgroup"
              />
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
